text/0208-ephemeral-resources.md

```diff
@@ -0,0 +1,78 @@
+
+Resources are managed via a ResourceProvider. Setting an attribute on a ResourceProvider will cause that attribute value to be included in the resource attached to any signal generated in the future. Spans which have already been started, along with any telemetry which has already been passed to the export pipeline, will not have the new attribute value. Optionally, a check can be added to ensure that permanent resources are not modified after the SDK has started
+
+
+## Internal details
+
+### ResourceProvider
+
+#### NewResourceProvider([resource], [validator]) ResourceProvider
+
+NewResourceProvider instantiates an implimentation of the ResourceProvider interface. As argumentes, it optionally takes an initial set of resource attributes, and a validator. 
+
+The ResourceProvider interface has the following methods
+
+#### MergeResource(resource)
+
+MergeResource creates a new resource, representing the union of the resource parameter and the resource contained within the Provider. The ResourceProvider holds a reference to the nwe resource.
+
+#### SetAttribute(key, value)
+
+SetAttribute functions the same as MergeResource, but only adds a single attribute.
+
+#### GetResource() Resource
+
+GetResource returns a reference to the current resource held by the ResourceProvider.
+
+#### FreezePermanent()
+
+FreezePermanent is called by the SDK one it has been stared. After FreezePermanent has been called, any calls to MergeResource or SetAttributes will only be applied if the validator acceptes the input.
+
+#### Implementation Notes
+
+For multithreaded systems, a lock should be used to queue all calls to MergeResource and SetAttribute. But the resource reference held by the ResourceProvider should be updated atomically, to prevent calls to GetResource from being blocked.
+
+### SDK Changes
+
+NewTraceProvider, NewMetricsProvider, and NewLogProvider now take a ResourceProvider as a parameter. For backwards compatibility, the Resource parameter remains functional. If both a resource and a resource provider as passed as parameters, the resource is merged into the ResourceProvider, then discarded.
+
+FreezePermanent is then called by the provider.
+
+Internally, providers hold a reference to the ResourceProvider, rather than a specific resource. When creating a signal, such as a span, metric, or log, GetResource() is called to obtain a reference to the correct resource to attach to the signal.
+
+## Trade-offs and mitigations
+
+This change should be fully backwards compatible, with one potential exception: fingerprinting. It is possible that an analysis tool which accepts OTLP may identify individual services by creating an identifier by hashing all of the resource attributes. 
+
+In this case, it is recommended that these systems modify their behavior, and choose a subset of permanent resources to use as a hash identifier.
```

Comment:
That might be a pretty big deal for some, if they only allow storing one set of resource attributes per hash.

Reply:
@open-telemetry/specs-approvers Please take a look - I suspect we may need a lot of eyes, in case somebody relies on this _right now_.

Reply:
It seems crazy to me to use a resource hash as an identifier, given that there is no requirement that the items within it would uniquely identify a service...

But I'm throwing it out there as a possibility, just to cover all the bases.

Reply:
You should be using something that doesn't exist yet, instead of hashing the whole resources: https://github.com/open-telemetry/opentelemetry-specification/issues/1034

Reply:
Yes I agree! There are various attributes which could count as a unique identifier. We could clarify in the spec which ones are currently defined. 

One possibility: by default, the SDK could generate a unique ID every time it starts, which would be a reliable identifier because we generate it ourselves. However, this identifier would not be stable across restarts. So there are limits to what can be provided without user input.